src/Analysis/Ast/Impl/Analyzer/ModuleWalker.cs

```diff
@@ -242,57 +242,103 @@ public ModuleWalker(IServiceContainer services, IPythonModule module, PythonAst
+                        foreach (var name in stubType.GetMemberNames().Except(new [] {"__class__", "__base__", "__bases__", "__mro__", "mro"})) {
+                            var stubMember = stubType.GetMember(name);
+                            var member = cls.GetMember(name);
+
+                            var memberType = member?.GetPythonType();
+                            var stubMemberType = stubMember.GetPythonType();
+
+                            if (builtins.Equals(memberType?.DeclaringModule) || builtins.Equals(stubMemberType?.DeclaringModule)) {
+                                continue; // Leave builtins alone.
+                            }
+
+                            if (!IsStubBetterType(memberType, stubMemberType)) {
+                                continue;
+                            }
+
+                            // Get documentation from the current type, if any, since stubs
+                            // typically do not contain documentation while scraped code does.
+                            TransferDocumentationAndLocation(memberType, stubMemberType);
+                            cls.AddMember(name, stubMember, overwrite: true);
+                        }
+                        break;
+
+                    case IPythonClassType _:
+                        // We do not re-declare classes, we only transfer members, see above.
+                        break;
+                    case IPythonModule _:
+                        // We do not re-declare modules.
+                        break;
+
+                    default:
+                        // Re-declare variable with the data from the stub unless member is a module.
+                        // Modules members that are modules should remain as they are, i.e. os.path
+                        // should remain library with its own stub attached.
+                        if (!(stubType is IPythonModule) && !builtins.Equals(stubType.DeclaringModule)) {
+                            TransferDocumentationAndLocation(sourceType, stubType);
+                            // TODO: choose best type between the scrape and the stub. Stub probably should always win.
+                            var source = Eval.CurrentScope.Variables[v.Name]?.Source ?? v.Source;
+                            Eval.DeclareVariable(v.Name, v.Value, source);
+                        }
+                        break;
                 }
             }
         }
 
         private static bool IsStubBetterType(IPythonType sourceType, IPythonType stubType) {
-            // If stub says 'Any' but we have better type, keep the current type.
             if (stubType.IsUnknown()) {
+                // Do not use worse types than what is in the module.
+                return false;
+            }
+            if (sourceType.IsUnknown()) {
+                return true; // Anything is better than unknowns.
+            }
+            if (sourceType.MemberType != stubType.MemberType) {
```

Comment:
Occasionally stub can have type of the same name but completely different type (like class vs function). We don't want to stop on the primary definitions.